tests/TestHelpers/HttpRequestHelper.php

```diff
@@ -144,36 +144,90 @@ public static function sendRequest(
-		} else {
-			$debugResponses = false;
+		// The exceptions that might happen here include:
+		// ConnectException - in that case there is no response. Don't catch the exception.
+		// RequestException - if there is something in the response then pass it back.
+		//                    otherwise re-throw the exception.
+		// GuzzleException - something else unexpected happened. Don't catch the exception.
+		try {
+			$response = $client->send($request);
+		} catch (RequestException $ex) {
+			$response = $ex->getResponse();
+
+			//if the response was null for some reason do not return it but re-throw
+			if ($response === null) {
+				throw $ex;
+			}
 		}
 
 		if ($debugRequests) {
 			self::debugRequest($request, $user, $password);
 		}
 
+		return $response;
+	}
+
+	/**
+	 *
+	 * @param string|null $url
+	 * @param string|null $xRequestId
+	 * @param string|null $method
+	 * @param string|null $user
+	 * @param string|null $password
+	 * @param array|null $headers ['X-MyHeader' => 'value']
+	 * @param mixed $body
+	 * @param array|null $config
+	 * @param CookieJar|null $cookies
+	 * @param bool $stream Set to true to stream a response rather
+	 *                     than download it all up-front.
+	 * @param int|null $timeout
+	 * @param Client|null $client
+	 *
+	 * @return ResponseInterface
+	 * @throws GuzzleException
+	 */
+	public static function sendRequest(
+		?string $url,
+		?string $xRequestId,
+		?string $method = 'GET',
+		?string $user = null,
+		?string $password = null,
+		?array $headers = null,
+		$body = null,
+		?array $config = null,
+		?CookieJar $cookies = null,
+		bool $stream = false,
+		?int $timeout = 0,
+		?Client $client =  null
+	):ResponseInterface {
+		if ((\getenv('DEBUG_ACCEPTANCE_RESPONSES') !== false) || (\getenv('DEBUG_ACCEPTANCE_API_CALLS') !== false)) {
+			$debugResponses = true;
+		} else {
+			$debugResponses = false;
+		}
+
 		$sendRetryLimit = self::numRetriesOnHttpTooEarly();
 		$sendCount = 0;
 		$sendExceptionHappened = false;
 
 		do {
-			// The exceptions that might happen here include:
-			// ConnectException - in that case there is no response. Don't catch the exception.
-			// RequestException - if there is something in the response then pass it back.
-			//                    otherwise re-throw the exception.
-			// GuzzleException - something else unexpected happened. Don't catch the exception.
-			try {
-				$response = $client->send($request);
-			} catch (RequestException $ex) {
-				$sendExceptionHappened = true;
-				$response = $ex->getResponse();
+			$response = self::sendRequestOnce(
+				$url,
+				$xRequestId,
+				$method,
+				$user,
+				$password,
+				$headers,
+				$body,
+				$config,
+				$cookies,
+				$stream,
+				$timeout,
+				$client
+			);
 
-				//if the response was null for some reason do not return it but re-throw
-				if ($response === null) {
-					throw $ex;
-				}
+			if ($response->getStatusCode() >= 400) {
+				$sendExceptionHappened = true;
```

Comment:
If the response status code is 425, then this code is going to set `$sendExceptionHappened = true`
And so `$loopAgain` will be false.
I think that this code will never loop again. So if there are random `425` "too early" that happen in ordinary test pipelines, the step will fail, when it should be retrying each second.

Reply:
Ohh!
I thought status >= 400 will be treated as `RequestException`, won't it be?

Reply:
It looks like the current logic is also broken if the above statement is correct.
We might need to wrap like this
```php
if ($response->getStatusCode() !== self::HTTP_TOO_EARLY) {
    $sendExceptionHappened = true;
}
```